sky/serve/serve_utils.py

```diff
@@ -414,21 +415,27 @@ def terminate_services(service_names: Optional[List[str]], purge: bool) -> str:
     for service_name in service_names:
         service_status = _get_service_status(service_name,
                                              with_replica_info=False)
-        assert service_status is not None, service_name
-        if service_status['status'] == serve_state.ServiceStatus.SHUTTING_DOWN:
+        if (service_status is not None and service_status['status']
+                == serve_state.ServiceStatus.SHUTTING_DOWN):
             # Already scheduled to be terminated.
             continue
-        if (service_status['status']
+        # If the `services` and `version_specs` table are not aligned, it might
+        # result in a None service status. In this case, the controller process
```

Comment:
Can we write down in what case the two tables will be misaligned if we have always handled the two database at the same time?

Reply:
Previously, the system added an initial version in the `__init__`` function of replica managers and added the service record here https://github.com/skypilot-org/skypilot/blob/9e27a81eaee897dc0c8b61a44a65d7b543eb527c/sky/serve/service.py#L146-L151

If the code is aborted before adding the version and after adding the service, it will cause a misalignment. In this PR we move the adding of initial version closer to the service table, to reduce the possibility of causing such misalignment, though arguably it is still 'possible' to happed, so I added a safe guard here.

Reply:
An example of abortion would be `shutil` lib call to fail, e.g. permission issues

Reply:
Was the previous issue mainly because the clean up not deleting the final version from the `version_spec` table, which is fixed by the `delete_all_versions` now?

If this safe guard is only for rare case, it would be great to mention that this is rare in the comment here to make sure we still know whether this is a common case in the future. : )

Reply:
Oh, sorry for not making it clear. The previous issue is the misalignment of `service` and `version_spec`, which causes the `_get_service_status` to return a `None` (due to the joining of the two tables), and causes the assertion to abort the deletion of the service.

https://github.com/skypilot-org/skypilot/blob/9e27a81eaee897dc0c8b61a44a65d7b543eb527c/sky/serve/serve_utils.py#L415-L417

I found that we did not clean up the remaining version when I debugged the original issue, and fix that in the same PR along the way.

Reply:
> If this safe guard is only for rare case, it would be great to mention that this is rare in the comment here to make sure we still know whether this is a common case in the future. : )

Good point! Added a comment for this. Thanks!